doc/dadbod.txt

```diff
@@ -136,7 +136,15 @@ Redis ~
 Redis doesn't have a username, so use a dummy one in the URL if a password is
 required.
 
-                                                *dadbod-sqlserver*
+                                                *dadbod-snowflake*
+Snowflake
+>
+    snowflake://[<accountname>.snowflakecomputing.com]/?[username=<username>&dbname=<dbname>&warehouse=<warehouse>]
```

Comment:
Absent a strong reason to otherwise, this should support the `user@host/dbname` syntax. Something like

```
snowflake://[<user>@][<accountname>][/<database>][?warehouse=<warehouse>]
```

Depending on how `warehouse works exactly, I could see doing something like one of the following:

```
snowflake://[<user>@][<accountname>]/[<database>[/<warehouse>]]
snowflake://[<user>@][<accountname>]/[<warehouse>[/<database>]]
```

The idea is you shouldn't have to remember if it's `user=` or `username=`, or `dbname=` or `database=`, just construct a URL like any other.

Reply:
Hi @tpope,
I am doing a lot of work in Snowflake ATM and looking at picking this back up as a Quality of Life improvement for myself. 

Would you be open a url format like
`snowflake:connection:[<warehouse>]/[<database>]/?options`, similar to the bigquery adapter?
Here, connection would be specified in the snowsql config file [as shown here](https://docs.snowflake.com/en/user-guide/snowsql-config)

Unfortunately, Snowflake does not support passing passwords as a CLI argument, it must be read in from this snowsql config file. Warehouse is a required field, it determines who is billed for any queries executed using the connection. I have it as optional here since it can be provided in the snowsql config as well.

There's a world where `snowflake://uname@account/...` could be supported by looking for the connection `username_account` in snowsql config, but this seems like it would be too restrictive/confusing to configure, since the actual username and account parameters would be overridden by whatever is in the snowsql config.

Reply:
> Hi @tpope, I am doing a lot of work in Snowflake ATM and looking at picking this back up as a Quality of Life improvement for myself.
> 
> Would you be open a url format like `snowflake:connection:[<warehouse>]/[<database>]/?options`, similar to the bigquery adapter? Here, connection would be specified in the snowsql config file [as shown here](https://docs.snowflake.com/en/user-guide/snowsql-config)

I think URLs should be supported as a baseline, but a shorthand for named connections could make sense. It's unclear to me if you're saying `connection` here is the name of a connection or the literal word "connection".

> Unfortunately, Snowflake does not support passing passwords as a CLI argument, it must be read in from this snowsql config file. Warehouse is a required field, it determines who is billed for any queries executed using the connection. I have it as optional here since it can be provided in the snowsql config as well.

Looking at [the docs](https://docs.snowflake.com/en/developer-guide/snowflake-cli-v2/connecting/specify-credentials), we could use `SNOWFLAKE_PASSWORD`. The SQL Server adapter demonstrates how to use environment variables.

> It seems to be possible: https://stackoverflow.com/questions/59203578/how-to-access-and-already-existing-snowflake-database-with-a-flask-application
> 
> ```
> 'snowflake://<user_login_name>:<password>@<account_name>/<database_name>/<schema_name>?warehouse=<warehouse_name>&role=<role_name>'
> ```

This looks perfect. It could be worth digging into how exactly that's implemented.

Reply:
> SNOWFLAKE_PASSWORD could work. One of my uses cases is to sign in to different environments in the same vim session, which use different accounts, so setting it even per-project is probably not going to cut it. I will dig a bit & come back.

It would be per URL. Either embed the password in the URL directly, or omit it and Dadbod will prompt for it and cache it for that URL exactly. It wouldn't reuse the password for a different user.

Or if you're using a connection from a config file, it's moot.

> The URL support from the stackoverflow post is in SQLAlchemy, which uses the `snowflake-connector-python` library (see [here](https://github.com/snowflakedb/snowflake-sqlalchemy/blob/58fb1bd556cddbd2b7cf2b8ccebf0ead3363e113/pyproject.toml#L41%5D)) which has sane URL support. This behavior is explicitly disallowed by their CLI tool SnowSQL. I could hack together a wrapper using the python connector, but would prefer not to when there is already an existing solution.

I don't think we should wrap it. I think we should copy what it does. I do not expect this to be super high effort. The `db#url#as_argv()` helper can translate URL properties into command line arguments.

The Python code that maps between properties of the URL (hostname, port, etc) and arguments to `snowflake.connector.connect()` is what we need. Then we just need to do the same thing but with command line arguments. Most of them are pretty obvious, but seeing the Python code would help confirm we aren't missing anything.

Reply:
> I think we should copy what it does.

I've been looking at this. I'm guessing that stackoverflow example would invoke SQLAlchemy's `create_engine` ([official docs](https://docs.sqlalchemy.org/en/20/core/connections.html)). The URL schema it uses is detailed in [this section](https://github.com/sqlalchemy/sqlalchemy/blob/077d058fad025f9a7edc4127ccd401c5e897935c/lib/sqlalchemy/engine/create.py#L119-L144). There's then [this place](https://github.com/sqlalchemy/sqlalchemy/blob/077d058fad025f9a7edc4127ccd401c5e897935c/lib/sqlalchemy/engine/interfaces.py#L1205) where the connection args are generated from the url, and from there I think we could trace down how to implement the snowflake connection. 


However, it seems that under the hood, the `snowflake-connector-python` library [doesn't](https://github.com/snowflakedb/snowflake-connector-python/blob/b05f25a3d72b18433d0f54fb34eb4cd96d9098f9/src/snowflake/connector/connection.py#L309) use the SnowSQL cli. Not sure if that's surprising, but it seems more straightforward to create an interface to the cli tool, reading connection parameters via a URL in the ways @tpope indicated above.'s

Reply:
> I've been looking into this. I'm guessing that stackoverflow example would create a SQLAlchemy `engine` ([official docs](https://docs.sqlalchemy.org/en/20/core/connections.html)). The URL schema that `create_engine` uses is detailed in [this section](https://github.com/sqlalchemy/sqlalchemy/blob/077d058fad025f9a7edc4127ccd401c5e897935c/lib/sqlalchemy/engine/create.py#L119-L144). There's then [this place](https://github.com/sqlalchemy/sqlalchemy/blob/077d058fad025f9a7edc4127ccd401c5e897935c/lib/sqlalchemy/engine/default.py#L631) where the connection args are generated from the url, and from there I think we could trace down how to implement the snowflake connection.

Right, you're most of the way there, just need to find the `url.translate_connect_args()` that's used for Snowflake.

Reply:
[Here we go!
](https://github.com/snowflakedb/snowflake-sqlalchemy/blob/58fb1bd556cddbd2b7cf2b8ccebf0ead3363e113/src/snowflake/sqlalchemy/snowdialect.py#L202)
(In snowflake-sqlalchemy, the actual SQLAlchemy dialect that wraps `snowflake-connector-python`)

Reply:
Apols for lack of attention, got a puppy this weekend and things have been hectic.
 
I have an implementation working with all but password using the snowsql cli. Is it worth bringing all this python code into this project as opposed to compromising on password provision and just documenting the required steps?

Reply:
> Is it worth bringing all this python code into this project as opposed to compromising on password provision and just documenting the required steps?

No need for either. Use `SNOWFLAKE_PASSWORD`, like this: https://github.com/tpope/vim-dadbod/blob/37523ffe0849b51db07fa7f48430c2aca5420a0e/autoload/db/adapter/sqlserver.vim#L46

Reply:
You have taught me a valuable vimscript lesson. Thank you! Pr tom